### PR TITLE
Add job abortion with tests

### LINKED ISSUE #76: Feature request: the ability to abort a job
Sometimes we may want to `abort` a job:
```javascript
const jobA = pool
  .run((input, done) => {
    somethingTakeALongTime()
    done(hash, input))
  .send({ do : 'something' })
// ...
jobA.abort()
```
I think that it can be implemented by adding a field `isAborted` in job object.
When user calls `job.abort()`,
1. If the job is executing on a thread, kill the thread and spawn a new one in the pool
2. If the job has not been executed, set the field `isAborted` to true, and add a judgement [here](https://github.com/andywer/threads.js/blob/634d4c1f33a2dfda304ec6b029b410170f438527/src/pool.js#L50) to check if a job is marked as aborted before execution.

### PR DESCRIPTION
**Pull request by Liu233w**
resolve #76 

I will add documents after the code being revewed.

### CODE REVIEW

**Comment by Liu233w** [comment]
I think that the reason why build failed is node10 changing it's api. see the [build log](https://travis-ci.org/andywer/threads.js/jobs/372073388#L2402)

File: src/pool.js [hunk 1]
```diff
@@ -46,11 +47,18 @@ export default class Pool extends EventEmitter {
     }
 
     const job = this.jobQueue.shift();
+
+    if (job.isAborted) {
```
Comment (andywer) [inline src/pool.js]: I would add a `job.once('abort', () => { /* unqueue job */ })` to `queueJob()` instead.

Comment (andywer) [reply src/pool.js]: Might need to add a new method: `dropJob(job)`

File: src/pool.js [hunk 2]
```diff
@@ -79,6 +87,18 @@ export default class Pool extends EventEmitter {
...
+  handleJobAbort(thread, job) {
+    thread.kill();
+
+    const index = this.threads.indexOf(thread);
+    this.threads.splice(index, 1);
+
+    const newThread = spawn(null, [], this.spawnOptions)
+    this.threads.push(newThread)
+
+    this.handleJobDone(newThread, job)
```
Comment (andywer) [inline src/pool.js]: `handleJobDone()`: Where is that method defined?

File: test/spec/pool.spec.js [hunk 3]
```diff
@@ -248,4 +264,95 @@ describe('Pool', () => {
...
+    before(() => {
+      fakeSpawn = RunnableFakeWorker;
+      Pool.spawn = origSpawn;
+    });
+
+    after(() => {
+      Pool.spawn = fakeSpawn;
+    });
+
+    it('can abort a job', (done) => {
```
Comment (andywer) [inline test/spec/pool.spec.js]: How about `can abort a running job`?

File: test/spec/pool.spec.js [hunk 4]
```diff
@@ -248,4 +264,95 @@ describe('Pool', () => {
...
+
+        // the original thread are not been deleted
+        expect(pool.threads[0]).to.equal(theThread)
+        expect(pool.idleThreads[0]).to.equal(theThread)
+
+        done();
+      }, 500);
+    });
+
+    it('can remove old thread and spawn new thread when a job is aborted', (done) => {
```
Comment (andywer) [inline test/spec/pool.spec.js]: 👍

File: src/pool.js (hunk 2 above)
Comment (Liu233w) [reply src/pool.js]: here: https://github.com/Liu233w/threads.js/blob/c5f067af2d3af7fed50f2c7c51d070f9adb65e93/src/pool.js#L80

**Comment by andywer** [comment]
Thanks for your contribution, @Liu233w! This is a nice feature :)

**Comment by Liu233w** [comment]
It's my pleasure.

By the way, I thought that if user added a listener on `abort`, it might be removed by [the code here](https://github.com/Liu233w/threads.js/blob/adc640a9fde1667174f3fd31a7b0205826d3adae/src/pool.js#L60). But I couldn't reproduce that case. I add a [test case](https://github.com/Liu233w/threads.js/blob/f8ce95bbcc012d1eb11370915a6f824ee90fc0d5/test/spec/pool.spec.js#L358) in a commit at temp branch.

**Comment by andywer** [comment]
Yeah, that is in fact a risk... I would merge this now, nevertheless.

But I think I just broke the build, somehow. Didn't change much: adc640a9fde1667174f3fd31a7b0205826d3adae

Tests succeed locally, too. Will try to re-run the build.

**Comment by andywer** [comment]
Ahh, the failing test is at least partially due to node 10.

**Comment by andywer** [comment]
#79 should fix the build. Just a bit concerned that the failed node 4 build is a heisenbug (failing non-deterministically from time to time)...

**Comment by andywer** [comment]
@Liu233w Can you please merge `master` into the feature branch? I tried to do it, but somehow it just fails:

```
andy@Andys-MacBook:~/workspace/threads.js$ git push Liu233w feat-abort-job
error: src refspec feat-abort-job does not match any.
error: failed to push some refs to 'https://github.com/Liu233w/threads.js.git'
```

**Comment by Liu233w** [comment]
ok. I'll rebase it to make it cleaner.

**Comment by andywer** [comment]
Published as `v0.11.0` 🚀

**Pull request merged**